quinn-proto/src/crypto.rs

```diff
@@ -186,3 +188,16 @@ pub trait HmacKey: Send + Sized + Sync {
     /// Method for verifying a message
     fn verify(&self, data: &[u8], signature: &[u8]) -> Result<(), ()>;
 }
+
+/// A key for sealing data with AEAD-based algorithms
+pub trait AeadKey {
+    /// Length of AEAD Key
+    const KEY_LEN: usize;
+
+    /// Method for deriving a fixed AEAD key using a `master_key` and `random_bytes`
+    fn derive_fixed(master_key: &[u8], random_bytes: &[u8]) -> Self;
```

Comment:
Let's name this `from_hkdf`, and accept a typed `HmacKey` rather than opaque key bytes. The `HmacKey` type can be obtained from a type parameter on `AeadKey`.

Reply:
I didn't fully understand this comment, could you please elaborate? 
It seems that you are suggesting the following,

```rust
    fn from_hkdf(master_key: HmacKey) -> Self; 
```

and then do an implementation 

```rust
impl crypto::AeadKey for ... {
    type HmacKey = hmac::Key;
    fn from_hkdf(master_key: HmacKey, random_bytes: &[u8]) -> Self {
       // How to derive using HmacKey instead of master_key bytes?
    }
}
```
Is that correct?

As as I understand from the docs of [`hkdf`](https://briansmith.org/rustdoc/ring/hkdf/index.html) module, we need the opaque key bytes so that we can build the output key material `Okm`.

Looking at the `hkdf` module APIs, I am not sure how `Okm` could be built from `hmac::Key`. 

In an earlier comment you said,

> We don't need two separate keys; after these changes token_key is unused. Rather than replacing token_key, I suggest using it as the master key directly.

Are you suggesting to combine `token_key.inner` and `token_key.outer` to build the IKM?

Reply:
> Looking at the hkdf module APIs, I am not sure how Okm could be built from hmac::Key.

Ah, you're right. What we should actually be using here is a(n abstraction over a) `ring::hkdf::Salt`; I was confusing `HmacKey` with that.

> Are you suggesting to combine token_key.inner and token_key.outer to build the IKM?

I'm not sure what `inner` and `outer` here are. My point is only that we only need one persistent key, and `token_key` can be what we call it.

Reply:
Er, rather, the `Prk` is the stable value we should be holding on to.

Reply:
> My point is only that we only need one persistent key, and token_key can be what we call it.

> Er, rather, the `Prk` is the stable value we should be holding on to.

Makes sense :+1: , then `token_key`  value would be a `Prk` which will be used to derive `Okm`, and that can be used to drive the `hmac` and `aead` keys :+1:

Reply:
Do we need an actual `hmac` key at all? The AEAD takes care of authentication on its own, and the `Prk` is sufficient for generating an AEAD key.

Reply:
> Do we need an actual hmac key at all? 

Thank you for putting this question, I didn't realize that the `token_key` wasn't used for any other place apart for generating message auth codes.